### DISCUSSION TITLE
How to take svg transformations into account when creating an element?

### DISCUSSION THREAD

**Discussion by Komar0ff** [clientIO/joint#2506] [Q&A] [answered]
### Introduction

Hello. I am trying to create a jointjs element with a 'markup' field. In it add svg:
```js
...
    const cell = new joint.dia.Element({
        position: { x: ..., y: ... },
        size: { width: ..., height: ... },
        markup: svgElement.outerHTML
        type: 'g-element',
    });
...
```
The svg itself has a transformation that mirrors the element horizontally.
The element is rendered correctly, but the selection is where the element would have been before the transformation
<img width="216" alt="Снимок экрана 2024-01-24 в 16 18 41" src="https://github.com/clientIO/joint/assets/10588170/0f09bb5c-e79d-4aaa-8ef7-1e5d4dc4aedf">

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS+

**Comment by Komar0ff** [comment]
As I understood the problem is in ui.FreeTransform which doesn't realize that a transformation has been applied to the element.
Is there any way to explicitly specify that you need to shift based on matrix transformations?

**Reply by kumilingus** [reply]
`ui.FreeTransform` works with model, not with the view It does not look at the SVG at all.

**Comment by Komar0ff** [comment]
Apparently it's easier to solve this problem by mutating the incoming svg element

**Reply by kumilingus** [reply]
You can also try following:
- adding your SVG to a `scalable` group (`<g @selector="scalable">${yourSVGString}</g>`)
- wrapping your SVG with group that negates the undesired transformation

**Reply by Komar0ff** [reply]
@kumilingus, thank you very much for your reply

> adding your SVG to a scalable group (<g @selector="scalable">${yourSVGString}</g>)

I apologize for the possibly stupid question. 
Where exactly do I add this line? 
If i add it to the markup field of an element, i will get a invalid XML error.

**Reply by kumilingus** [reply]
How do you use your SVG markup now?

```js
dia.Element.define('MyElement', {}, {
  markup: util.svg`<g @selector="scalable"><g transform="matrix(...)" id="1.0">...</g></g>`
});
```

**Reply by Komar0ff** [reply]
@kumilingus I've just put together some sample code that reproduces my problem.
https://github.com/Komar0ff/react-joint/tree/main

[Adding](https://github.com/Komar0ff/react-joint/blob/main/src/App.tsx#L35) @selector="scalable" unfortunately did not solve the problem(

**Reply by kumilingus** [reply]
I see. The `scalable` group only scales its content to fit the model size. You'd also have to reset the `translation`.

```js
const svgString = `<g transform="matrix(1, 0, 0, -1, 0, 0)" id="33.0">     <title>Terminator</title>     <userdefs>      <ud nameu="visNavOrder" val="33000"></ud>      <ud nameu="visVersion" val="15"></ud>     </userdefs>     <custprops>      <cp nameu="Cost" lbl="Cost" type="7.0" langid="1033.0" val="0"></cp>      <cp nameu="Duration" lbl="Duration" type="2.0" langid="1033.0" val="0"></cp>      <cp nameu="Resources" lbl="Resources" type="0.0" langid="1033.0" val="0"></cp>     </custprops>     <path d="M13.5,0 L77.590774536,0 L77.590774536,0 C81.169815063,0 84.605941772,1.423288465 87.136711121,3.954058886 L87.136711121,3.954058886 C89.667480469,6.484828949 91.090774536,9.920949936 91.090774536,13.5 L91.090774536,13.5 C91.090774536,17.079050064 89.667480469,20.515171051 87.136711121,23.045940399 L87.136711121,23.045940399 C84.605941772,25.576713562 81.169815063,27 77.590774536,27 L77.590774536,27 L13.5,27 L13.5,27 C9.920949936,27 6.484828949,25.576713562 3.954058886,23.045940399 L3.954058886,23.045940399 C1.423288465,20.515171051 0,17.079050064 0,13.5 L0,13.5 C0,9.920949936 1.423288465,6.484828949 3.954058886,3.954058886 L3.954058886,3.954058886 C6.484828949,1.423288465 9.920949936,0 13.5,0 Z" stroke="#316176" stroke-width="1.5" stroke-linecap="round" fill="#D8CBB3" fill-rule="evenodd"></path>     <g transform="matrix(1.0,0.0,0.0,1.0,0.0,0.0)">      <text x="40.550266" y="18.0" class="TextStyle1" fill="#3F3F3F" opacity="1.0" transform="matrix(1.0,0.0,0.0,-1.0,0.0,22.5)">-</text>     </g>    </g>`;

const el1 = new dia.Element({
  position: { x: 100, y: 100 },
  size: { width: 100, height: 50 },
  attrs: {
    reset: {
        ref: 'scalable',
        transform: 'translate(calc(-1*x), calc(-1*y))'
    }
  },
  markup: util.svg`<g @selector="reset"><g @selector="scalable">${svgString}</g></g>`,
  type: 'custom-element'
});

el1.addTo(graph)
```

Note that you could also use the `svgString` to construct a data URL and use it as `href` attribute of an `<image/>`.
See https://www.jointjs.com/demos/external-svg-images

**Reply by Komar0ff** [reply]
@kumilingus 
> You'd also have to reset the translation

Unfortunately, it is not as easy to do as in the presented example. 
In real life matrix calculations can be more complicated. For example:
`matrix(-0.5379140973091125, 0.8429996371269226, 0.8429996371269226, -1.5379140973091125, 0, 0)`
At the moment I have no idea how to explicitly reset such calculations 😓

> Note that you could also use the svgString to construct a data URL and use it as href attribute of an <image/>

And in this case how to work with text blocks that are also transformed? After all, if they are like an image, we can't change the text for them. And unfortunately this method will not solve the matrix transformation problem either(

**Reply by kumilingus** [reply]
It does not matter what is the `matrix` transformation. It will always reset it so it fits the model bounding box.

```js
const svgString = `<g transform="matrix(-0.5379140973091125, 0.8429996371269226, 0.8429996371269226, -1.5379140973091125, 0, 0)" id="33.0">     <title>Terminator</title>     <userdefs>      <ud nameu="visNavOrder" val="33000"></ud>      <ud nameu="visVersion" val="15"></ud>     </userdefs>     <custprops>      <cp nameu="Cost" lbl="Cost" type="7.0" langid="1033.0" val="0"></cp>      <cp nameu="Duration" lbl="Duration" type="2.0" langid="1033.0" val="0"></cp>      <cp nameu="Resources" lbl="Resources" type="0.0" langid="1033.0" val="0"></cp>     </custprops>     <path d="M13.5,0 L77.590774536,0 L77.590774536,0 C81.169815063,0 84.605941772,1.423288465 87.136711121,3.954058886 L87.136711121,3.954058886 C89.667480469,6.484828949 91.090774536,9.920949936 91.090774536,13.5 L91.090774536,13.5 C91.090774536,17.079050064 89.667480469,20.515171051 87.136711121,23.045940399 L87.136711121,23.045940399 C84.605941772,25.576713562 81.169815063,27 77.590774536,27 L77.590774536,27 L13.5,27 L13.5,27 C9.920949936,27 6.484828949,25.576713562 3.954058886,23.045940399 L3.954058886,23.045940399 C1.423288465,20.515171051 0,17.079050064 0,13.5 L0,13.5 C0,9.920949936 1.423288465,6.484828949 3.954058886,3.954058886 L3.954058886,3.954058886 C6.484828949,1.423288465 9.920949936,0 13.5,0 Z" stroke="#316176" stroke-width="1.5" stroke-linecap="round" fill="#D8CBB3" fill-rule="evenodd"></path>     <g transform="matrix(1.0,0.0,0.0,1.0,0.0,0.0)">      <text x="40.550266" y="18.0" class="TextStyle1" fill="#3F3F3F" opacity="1.0" transform="matrix(1.0,0.0,0.0,-1.0,0.0,22.5)">-</text>     </g>    </g>`;

const el1 = new dia.Element({
  position: { x: 100, y: 100 },
  size: { width: 100, height: 50 },
  attrs: {
    resetTransform: {
        ref: 'scalable',
        transform: 'translate(calc(-1*x), calc(-1*y))'
    },
    bg: {
        stroke: 'red',
        fill: 'transparent',
        width: 'calc(w)',
        height: 'calc(h)',
    }
  },
  markup: util.svg`<rect @selector="bg"></rect><g @selector="resetTransform"><g @selector="scalable">${svgString}</g></g>`,
  type: 'custom-element'
});

el1.addTo(graph)
```

> And in this case how to work with text blocks that are also transformed? After all, if they are like an image, we can't change the text for them. And unfortunately this method will not solve the matrix transformation problem either(

But if you allow the elements to be resized, the text with the provided SVG will be skewed anyway if you are not preserving the original aspect ratio.

**Reply by Komar0ff** [reply]
Hmm, the highlighting is indeed fixed. 
But unfortunately the schematics are now visually broken.
Before:
<img width="371" alt="Снимок экрана 2024-02-12 в 12 52 41" src="https://github.com/clientIO/joint/assets/10588170/faacca6c-cdf0-49d7-bdd0-7e6c4f38efd0">

After:
<img width="415" alt="Снимок экрана 2024-02-12 в 12 31 45" src="https://github.com/clientIO/joint/assets/10588170/ca9beb27-4334-440c-a3c7-890ac440ae7c">

The svg of the connection lines looks something like this:

```xml
<g transform='matrix(1.0,0.0,0.0,-1,0,0)'>
  <title>Dynamic connector.37</title>
  <path d='M126.869659424,4.206614017 L132.341674805,7.086614132 L126.869659424,9.966614723 Z' stroke='#316176' stroke-width='1.5' fill='#316176'  fill-rule='evenodd' transform='matrix(-0.0,-1.0,1.0,0.0,127.559044,141.73227)'/>
  <path d='M0,7.086614132 L0,17.716535568 L0,17.716535568 L134.6456604,17.716535568 L134.6456604,17.716535568 L134.6456604,9.966614723 ' stroke='#316176' stroke-width='1.5' stroke-linecap='round' fill='none'/>
</g>

```

**Reply by kumilingus** [reply]
You connectors are also elements, correct?
Perhaps you should not reset the transformation for the connectors.

**Reply by Komar0ff** [reply]
@kumilingus 
Yeah, that's right.
Unfortunately, if you do not apply reset transformation again there are problems with highlight
<img width="215" alt="Снимок экрана 2024-02-12 в 13 58 21" src="https://github.com/clientIO/joint/assets/10588170/9765bf72-ee90-4849-bab3-aa2a4ab6065c">

**Reply by Komar0ff** [reply]
@kumilingus 
It's probably worth talking about the context here:
Unfortunately, the tool I use to convert other file formats first creates an inverted svg and then flips each element individually (hence so many matrix transformations).
Hence, I need to compute the matrix transformations for each element and generate a joint cell.
For now, I consider an element to be a top-level and non-empty <g> tag. 

```js
function combineTransforms(parentMatrix, componentMatrix) {

    const { a: a1, b: b1, c: c1, d: d1, e: e1, f: f1 } = parentMatrix
    const { a: a2, b: b2, c: c2, d: d2, e: e2, f: f2 } = componentMatrix

    return {
        // scaleX
        a: a1 * a2 + c1 * b2,
        // skewY
        b: b1 * a2 + d1 * b2,
        // skewX
        c: a1 * c2 + c1 * d2,
        // scaleY
        d: b1 * c2 + d1 * d2,
        // translateX
        e: a1 * e2 + c1 * f2 + e1,
        // translateY
        f: b1 * e2 + d1 * f2 + f1,
    }
}
```
Then determining the exact location of the item and its transformations (whatever they are). I would like the selection to match the position of the element. 
```js
    ....
    if (parentMatrix && componentMatrix) {
        mutateMatrix = combineTransforms(parentMatrix, componentMatrix)
        sanitizeElement = element.outerHTML.replace(
            /transform=".*?"/,
            `transform="matrix(${mutateMatrix.a}, ${mutateMatrix.b}, ${mutateMatrix.c}, ${mutateMatrix.d}, 0, 0)"`
        )
    }

    const cell = new dia.Element({
        position: {
            x: componentMatrix.e,
            y: componentMatrix.f,
        },
        attrs: {
          resetTransform: {
            ref: 'scalable',
            transform: 'translate(calc(-1*x), calc(-1*y))'
          }
        },
        size: { width: GET_BOX.width, height: GET_BOX.height },
        markup: util.svg`<g @selector="resetTransform"><g @selector='scalable'>${sanitizeElement}</g></g>`,
        type: 'g-element',
    })

    return cell

```
Perhaps I am somewhere wrong in my decisions in forming the joint cell?

**Reply by kumilingus** [reply]
> For now, I consider an element to be a top-level and non-empty tag.

Does it mean that the `parentMatrix` computation is not relevant for this diagram?
<img width="371" alt="Снимок экрана 2024-02-12 в 12 52 41" src="https://github.com/clientIO/joint/assets/10588170/faacca6c-cdf0-49d7-bdd0-7e6c4f38efd0">

Where does the `componentMatrix` come from?

**Reply by Komar0ff** [reply]
@kumilingus 
> Does it mean that the parentMatrix computation is not relevant for this diagram?

Good question. The general structure of the svg file after conversion is as follows:
```xml
// xml boilerplate
...
<g id='SFixTitle'/>
 <g id='SContent'>
  <g>
   <g transform='matrix(...)'>
     // Here are all the <g> tag elements that will be converted to joint cells
   </g>
  </g>
</g>
```
This means I need to always consider the matrix transformations of the parent. 

To put it simply: I take an SVG and make it flat taking into account matrix transformations of descendants, and then I turn each element of this flat structure into a joint cell.
If we specifically talk about the diagram above, its structure looks as follows: 
![tmp-0 2606168597950347](https://github.com/clientIO/joint/assets/10588170/159247fd-bbab-404a-a500-db9732f09776)


> Where does the componentMatrix come from?

componentMatrix i take from the current element:
```js
    const componentMatrix = element.transform.baseVal.consolidate().matrix
```

**Reply by Komar0ff** [reply]
@kumilingus Any ideas on how this can be solved?

**Reply by kumilingus** [reply]
We've done the same thing with MS Visio. There is most likely an issue with the transformation logic.

---

Could you perhaps create an example e.g. diagram above, that can I run? (with inverted SVG on input)?

You can also start without reverting the `y`  coordinate.
It looks to me that the SVG is flipped with `matrix(1, 0, 0, -1, 0, 0)` but it must be also translated with `matrix(1, 0, 0, -1, 0, -HEIGHT)`.

![image](https://github.com/clientIO/joint/assets/3967880/eb4f1774-a3ef-40eb-b56d-18b1dbe2906d)

---

> `const componentMatrix = element.transform.baseVal.consolidate().matrix`

What `element` is this? The JointJS element  is not rendered yet.

**Reply by Komar0ff** [reply]
@kumilingus 
I put together a small example reproducing my problem with different diagrams. https://github.com/Komar0ff/react-joint.


> What element is this? The JointJS element is not rendered yet.

`element` is an element that I get after implicitly adding svg to the DOM for more accurate sizing of the element (temporary solution). https://github.com/Komar0ff/react-joint/blob/main/src/convertToJointCells.js#L9

**Reply by Komar0ff** [reply]
@kumilingus

>You can also start without reverting the y coordinate.
It looks to me that the SVG is flipped with matrix(1, 0, 0, -1, 0, 0) but it must be also translated with matrix(1, 0, 0, -1, 0, -HEIGHT)

Tried this method, but unfortunately it also did not solve my problem( Any other ideas?

**Reply by kumilingus** [reply, marked as answer]
It seems like the connectors and nodes require different logic (not sure how you converted them in the first place).

> You can also start without reverting the y coordinate.
It looks to me that the SVG is flipped with matrix(1, 0, 0, -1, 0, 0) but it must be also translated with matrix(1, 0, 0, -1, 0, -HEIGHT)
Tried this method, but unfortunately it also did not solve my problem( Any other ideas?

It actually quite worked for me.

```js
const generateComplexCell = (element, parentMatrix) => {
    const GET_BOX = element.getBBox()

    const componentMatrix = element.transform.baseVal.consolidate().matrix
    let sanitizeElement = null
    let mutateMatrix = null

    if (parentMatrix && componentMatrix) {
        mutateMatrix = combineTransforms(parentMatrix, componentMatrix)
        sanitizeElement = element.outerHTML.replace(
            /transform=".*?"/,
            `transform="matrix(${mutateMatrix.a}, ${mutateMatrix.b}, ${mutateMatrix.c}, ${mutateMatrix.d}, 0, ${GET_BOX.height})"`
        )
    }

    const cell = new dia.Element({
        position: {
            x: mutateMatrix?.e || componentMatrix.e,
            y: (mutateMatrix?.f || componentMatrix.f) - GET_BOX.height,
        },
        size: { width: GET_BOX.width, height: GET_BOX.height },
        markup: sanitizeElement || element.outerHTML,
        type: 'g-element',
    })

    cell.attr('root/title', 'autograph.Element')
    return cell
}
```

**Reply by Komar0ff** [reply]
@kumilingus 
Thank you so much for your responses. It really works for the elements. Also you gave me an idea how to do it for text and connecting lines as well. 
Really appreciate what you are doing. Thank you
